### ISSUE TITLE
[CLOSED] installer does not replace keralua.dll if installing over existing Helios

### ISSUE THREAD

**Issue by BlueFinBima** [BlueFinBima/test1#204]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [derammo](https://github.com/derammo)**
_Thursday Mar 12, 2020 at 01:54 GMT_
_Originally opened as https://github.com/BlueFinBima/Helios/issues/204_

----

Working with a user who sees Helios crash every time they try to run 1.4.2020.0226

system log:

```
- <Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event%22%3E
- <System>
  <Provider Name=".NET Runtime" /> 
  <EventID Qualifiers="0">1026</EventID> 
  <Level>2</Level> 
  <Task>0</Task> 
  <Keywords>0x80000000000000</Keywords> 
  <TimeCreated SystemTime="2020-03-10T04:50:50.241211400Z" /> 
  <EventRecordID>25502</EventRecordID> 
  <Channel>Application</Channel> 
  <Computer>XXXXXXXXXXX</Computer> 
  <Security /> 
  </System>
- <EventData>
  <Data>Anwendung: Control Center.exe Frameworkversion: v4.0.30319 Beschreibung: Der Prozess wurde aufgrund einer unbehandelten Ausnahme beendet. Ausnahmeinformationen: System.IO.FileLoadException bei NLua.Lua.Close() bei NLua.Lua.Dispose() bei NLua.Lua.Finalize()</Data> 
  </EventData>
  </Event>
```

interpretation:  

- It appears to be saying it is running on Framework 4.0.30319 (which is the RTM version of .NET 4.0) 
- the exception means "can't load assembly even though I found it"
- the exception trace (Finalize -> Dispose -> Close) makes no sense at all but I assume it is failing dramatically and getting a bad trace as a result

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Thursday Mar 12, 2020 at 01:55 GMT_

----

https://docs.microsoft.com/en-us/dotnet/api/system.io.fileloadexception?view=netframework-4.8
https://social.msdn.microsoft.com/Forums/en-US/fa4776a1-95fd-405d-9098-a89e37f85d9e/where-to-download-net-framework-v4030319?forum=netfxsetup
https://github.com/NLua/NLua/blob/v1.4.29/src/Lua.cs

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Thursday Mar 12, 2020 at 03:42 GMT_

----

ok I need more information / knowledge.   I just created a virtual machine, installed plain win10 pro on it, installed Helios, ran the F16C profile that the user was saying was crashing, and it worked fine.  I fully expected Helios to fail to run because of a missing Framework, but it did not.  In the documented install locations for .NET framework, all I find is these:

![image](https://user-images.githubusercontent.com/817160/76485012-a1244b80-63f1-11ea-8ea9-ab570c1a870a.png)

![image](https://user-images.githubusercontent.com/817160/76485032-b13c2b00-63f1-11ea-9d0a-b783a446abf4.png)

Obviously, I don't have enough information about how this is supposed to work, because this makes no sense.  We declare 4.5 as our minimum, so how are these satisfied?   Why is it working?   There are a bunch of separate assemblies in the "assemblies" folder.  Does it just satisfy individual references from there and not actually use the frameworks folders?

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Thursday Mar 12, 2020 at 03:49 GMT_

----

here is the answer:

All versions of .NET 4 are installed in the same path (4.0.30319) because they are installed as upgrades of that.   How do these MS developers sleep at night?  Anyway, the real version info is here in the registry:

![image](https://user-images.githubusercontent.com/817160/76485423-f01eb080-63f2-11ea-81e4-88d094d20251.png)

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Thursday Mar 12, 2020 at 03:59 GMT_

----

and yes that means that a fresh windows 10 (pro, 64 bit) install has .NET Framework 4.7.x installed already.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Thursday Mar 12, 2020 at 06:17 GMT_

----

@derammo Is the person reporting the problem running Win10?

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Thursday Mar 12, 2020 at 11:56 GMT_

----

no, MacOS.   But I decided to repro with Win10 anyway 🤣   ok not seriously, but i’m waiting for confirmation they aren’t on 32 bit

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Thursday Mar 12, 2020 at 12:00 GMT_

----

they have 64 bit win10 with .net 4.8 runnning in German

i will repro that config and keep working on it

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Thursday Mar 12, 2020 at 13:24 GMT_

----

Pretty sure I am only running 4.7.  When the dust settles, I wonder if this
will turn out to be the same issue I was looking at last weekend which had
no symptoms other than failing to start.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Thursday Mar 12, 2020 at 18:21 GMT_

----

this is all red herrings.  Like I suggested above, the trace is nonsense.  This problem happened because installing 2020.226 over the 2019.1005 release did not replace keralua.dll.   The version number in the dll seems to have gone DOWN so the installer didn't replace it?   At least I think so.  Will repro when I am off duty.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 02:22 GMT_

----

confirmed with .226 over .1005, keralua isn't replaced and it crashed on run

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 02:28 GMT_

----

confirmed with .308 over .1005, same problem
for the record, it does log in ControlCenter.log:
```
3/12/2020 7:26:32 PM - Error - Binding threw unhandled exception. (Binding="when NetWork Values NetWorkValue_10 on DCS Generic changes set angle of attack on AOA to  trigger value")
Exception Source:NLua
Exception Message:Could not load file or assembly 'KeraLua, Version=1.0.26.0, Culture=neutral, PublicKeyToken=6a194c04b9c89217' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
Stack Trace:
   at NLua.Lua..ctor()
   at GadrocsWorkshop.Helios.HeliosBinding.OnTriggerFired(Object trigger, HeliosTriggerEventArgs e)
```

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 02:52 GMT_

----

Ok stuck now.  Full installer products like InstallShield or WiX have an option "Always Overwrite" which basically just sets the version of the file in the .msi table to 65535.x.x.x so it will always be considered newer and will overwrite whatever.   The file itself isn't stamped with that version, it is just changed in the table.
https://helpnet.flexerasoftware.com/installshield23helplib/helplibrary/IDlgFileProp.htm#ref-dialogboxesa-m_2762307071_wp1157

This means we need to do that sort of thing in our post build script to update the file versions of third party libraries to not be taken from the actual DLLs but instead either set it to infinity or just set all of them to our product version.   Either way, it apparently can't be done cleanly in the limited Installer Project interface that we have, without hackery.

I think we need to disable in-place upgrades in our MSI so it forces people to remove Helios first, unless we can quickly resolve this problem with post build scripting changes to the MSI tables.  I don't think we can change the version numbers on the files because that would break dependencies between assemblies.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 02:57 GMT_

----

summon @BlueFinBima

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 02:59 GMT_

----

(and of course we need to tell folks to uninstall previous Helios before installing any 1.4.2020 that we have already released)

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 03:28 GMT_

----

in one test my binary diff also showed that phidgets.dll was kept from previous version, presumably the version number hasn't been bumped.  However, the binaries were different so code generation / platform changes must be checked and the version number bumped if anything is changed.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Friday Mar 13, 2020 at 12:18 GMT_

----

@derammo I am very confused.  My Windows 10 does not allow me to install
without uninstalling the previous version.  What magic are you doing to
coax the installer to slap the new over the top of the previous?

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 12:35 GMT_

----

nothing.   install 2019.1005,  install vc redist to make it run,  test run,  install 2020.x

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 12:47 GMT_

----

presumably, you have some sort of artifact left over from a previous install.   I recommend you try uninstall Helios, wipe the install folder, and then try again from installing 2019.1005 fresh.   If it still doesn't overwrite, then study the technical info about what prevents upgrades in msi.   There is stuff like version files and such, maybe there is something in the registry or in the installed software database that needs cleaning.  I haven't done the research.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Friday Mar 13, 2020 at 16:03 GMT_

----

The file version numbering of KeraLua.dll looks **VERY** suspicious with the 2014 version reporting 1.3.2.0 and the current file being 1.0.26.0.  The older file has a higher version number!  The installer might be flakey, but I'm going to let it off with not updating KeraLua.dll in this instance.  I don't want to perpetually force an uninstall of the previous dll, Will need to work something out.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 18:16 GMT_

----

no the old one was 1.3.x.    that’s why it won’t replace it with something called 1.0.26

did you actually install 1005 or something else?

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Friday Mar 13, 2020 at 18:24 GMT_

----

I had edited that entry out of all recognition to the original post.  I
can't work out how to do it the proper way, but I can do it in a way that
will simply confirm that I am a complete cowboy.  "Yee haa" as you
Americans say.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 18:25 GMT_

----

no worries.   getting second guessed by someone who is wrong is my absolute favorite thing in the whole world

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Friday Mar 13, 2020 at 21:33 GMT_

----

Actually looking back in the discourse, the fix I came up with is remarkably similar to the technique you described being used by other packages so maybe it isn't such a kludge (I'm kidding myself, it is still a kludge).  https://github.com/BlueFinBima/Helios/commit/b9dad44ac747ad9287de2e54d9210eca213afec8

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Friday Mar 13, 2020 at 21:34 GMT_

----

I'll probably package this up tomorrow if time allows.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Friday Mar 13, 2020 at 21:48 GMT_

----

ok feel free to assign to you then.  it was assigned to me so i assumed i was implementing it, but happy to have it off my plate.

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Tuesday Mar 17, 2020 at 12:19 GMT_

----

Fixed by a combination of activities in 1.4.2020.0315 .
